File: src/ipc/history.ts
import { invoke } from "@/ipc/invoke";
import type { DocumentId } from "@/ipc/pdf";

/**
 * Undo/redo availability for a document. Mirrors
 * `pdf::undo::HistoryState` on the Rust side. The frontend stores this
 * per document to drive the Undo/Redo button state; the actual stacks
 * live in the document actor (the frontend never edits PDF bytes).
 */
export interface HistoryState {
  canUndo: boolean;
  canRedo: boolean;
}

/** SPEC: P2-PAGE-003 / session history — undo the most recent edit. */
export async function undo(id: DocumentId): Promise<HistoryState> {
  return invoke<HistoryState>("pdf_undo", { id });
}

/** Redo the most recently undone edit. */
export async function redo(id: DocumentId): Promise<HistoryState> {
  return invoke<HistoryState>("pdf_redo", { id });
}

/** Query current availability (used to hydrate the store on open). */
export async function historyState(id: DocumentId): Promise<HistoryState> {
  return invoke<HistoryState>("pdf_history_state", { id });
}

File: src-tauri/src/commands/history.rs
use tauri::State;

use crate::error::CommandError;
use crate::pdf::undo::HistoryState;
use crate::AppState;

/// Resolve `id` to an actor and run `send` (a send-only request method on
/// the handle) while holding the map lock, returning the reply receiver.
/// The lock is dropped before the caller awaits — same discipline as
/// `pdf_render_page` / `pdf_save`.
macro_rules! request {
    ($state:expr, $id:expr, $method:ident) => {{
        let uuid = uuid::Uuid::parse_str(&$id)
            .map_err(|_| CommandError::InvalidInput(format!("not a UUID: {}", $id)))?;
        let guard = $state
            .actors
            .lock()
            .map_err(|e| CommandError::Internal(format!("actor map poisoned: {e}")))?;
        let handle = guard
            .get(&uuid)
            .ok_or_else(|| CommandError::NotFound(format!("document {}", $id)))?;
        handle.$method()?
    }};
}

/// SPEC: P2-PAGE-003 / session history — undo the most recent edit.
/// A no-op (returns the unchanged availability) when nothing is undoable.
#[tauri::command]
pub async fn pdf_undo(
    state: State<'_, AppState>,
    id: String,
) -> Result<HistoryState, CommandError> {
    let rx = request!(state, id, undo_request);
    rx.await
        .map_err(|_| CommandError::Internal("doc-actor dropped reply".into()))?
}

/// Redo the most recently undone edit. No-op when nothing is redoable.
#[tauri::command]
pub async fn pdf_redo(
    state: State<'_, AppState>,
    id: String,
) -> Result<HistoryState, CommandError> {
    let rx = request!(state, id, redo_request);
    rx.await
        .map_err(|_| CommandError::Internal("doc-actor dropped reply".into()))?
}

/// Current undo/redo availability, for hydrating the UI button state.
#[tauri::command]
pub async fn pdf_history_state(
    state: State<'_, AppState>,
    id: String,
) -> Result<HistoryState, CommandError> {
    let rx = request!(state, id, history_state_request);
    rx.await
        .map_err(|_| CommandError::Internal("doc-actor dropped reply".into()))
}
